File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    int f(vector<int>& v, int i, int j) {
        if (i == j) return v[i];
        
        int l = v[i] - f(v, i + 1, j);
        int r = v[j] - f(v, i, j - 1);
        
        return max(l, r);
    }
    
    bool predictTheWinner(vector<int>& nums) {
        return f(nums, 0, nums.size() - 1) >= 0;
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:
        def f(i,j):
            if i == j:
                return nums[i]

            l = nums[i] - f(i+1, j)
            r = nums[j] - f(i, j-1)

            return max(l,r)

        return f(0, len(nums) - 1) >= 0            
        